great_ape_safe/ape_api/balancer.py

```diff
@@ -152,47 +210,140 @@ def unstake_all_and_withdraw_all(
             pool_id = pool.getPoolId()
 
         underlyings, reserves, _ = self.vault.getPoolTokens(pool_id)
+        gauge = self.safe.contract(self.gauge_factory.getPoolGauge(pool))
 
-        if unstake:
-            balance_before = pool.balanceOf(self.safe)
-
-            gauge = self.safe.contract(self.gauge_factory.getPoolGauge(pool))
-            gauge.withdraw(gauge.balanceOf(self.safe))
+        amount_in = pool.balanceOf(self.safe)
 
-            assert pool.balanceOf(self.safe) > balance_before
+        if unstake:
+            amount_in += gauge.balanceOf(self.safe)
 
-        exact_bpt_for_tokens_enum = 1
-        amount_in = pool.balanceOf(self.safe)
 
         data_encoded = eth_abi.encode_abi(
             ['uint256', 'uint256'],
-            [exact_bpt_for_tokens_enum, amount_in]
+            [1, amount_in]
             )
 
-        underlyings_out = calc_tokens_out_given_exact_bpt_in(
+        is_stable = pool.symbol()[0:3] == 'sta'
+
+        if is_stable:
+            underlyings_out = StableMath.calcTokensOutGivenExactBptIn(
+                pool, reserves, amount_in
+            )
+        else:
+            underlyings_out = WeightedMath.calc_tokens_out_given_exact_bpt_in(
             pool, reserves, amount_in
         )
+
         min_underlyings_out = \
             [x * (1 - self.max_slippage) for x in underlyings_out]
 
+        if is_eth:
+            underlyings = list(underlyings)
+            weth_index = underlyings.index(self.weth)
+            underlyings[weth_index] = ZERO_ADDRESS
+
         request = (
             underlyings,
             min_underlyings_out,
             data_encoded,
             is_eth
         )
 
-        balances_before = [Contract(x).balanceOf(destination) for x in underlyings]
+        pool.approve(self.vault, amount_in)
+
+        self._unstake_and_withdraw_all(
+            pool, request, unstake, claim, destination
+        )
+
+
+    def unstake_and_withdraw_all_single_asset(
+        self, asset, underlyings=None, pool=None, unstake=True, claim=True, is_eth=False, destination=None
+    ):
+        if not underlyings and not pool:
+            raise TypeError('must provide either underlyings or pool')
+
+        if underlyings:
+            underlyings = sorted(underlyings)
```

Comment:
```suggestion
            underlyings = sorted([x.address for x in underlyings])
```
also in L206 but i cant comment there for some reason